thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/DetectionPipeline.java

```diff
@@ -90,22 +90,21 @@ private void initComponents() throws Exception {
     Map<String, BaseComponent> instancesMap = config.getComponents();
     Map<String, Object> componentSpecs = config.getComponentSpecs();
     if (componentSpecs != null) {
-      for (String componentName : componentSpecs.keySet()) {
-        Map<String, Object> componentSpec = MapUtils.getMap(componentSpecs, componentName);
-        if (!instancesMap.containsKey(componentName)){
-          instancesMap.put(componentName, createComponent(componentSpec));
+      for (String componentKey : componentSpecs.keySet()) {
+        Map<String, Object> componentSpec = MapUtils.getMap(componentSpecs, componentKey);
+        if (!instancesMap.containsKey(componentKey)){
+          instancesMap.put(componentKey, createComponent(componentSpec));
         }
       }
 
-      for (String componentName : componentSpecs.keySet()) {
-        Map<String, Object> componentSpec = MapUtils.getMap(componentSpecs, componentName);
+      for (String componentKey : componentSpecs.keySet()) {
```

Comment:
Can we combine the two for loops?

Reply:
Not really! In the second loop, we replace all component references with the components. This comes by looking up the componentKey to Component mapping which is generated in the first loop.

Reply:
You can do  update right after inserting.
for (String componentKey : componentSpecs.keySet()) {
// check if key exists
// if not then insert (key, value)
// update the map (value)
}

Reply:
Sorry, did I miss something! Are you talking about merging the 2 sequential loops on `componentSpecs` or the 2 nested loops?

My earlier comment was addressed assuming the former case. I am still not convinced even if it is the latter case. Note that the two loops are on `componentSpecs` and `componentSpec`. The parent loop iterates over each component and the nested loop iterates over the specifications of each component.

I do not have a complete understanding of the reference logic here either. Nevertheless, if required I will make this change in a separate PR as it is unrelated to this change.

Reply:
Sure. Thanks for the details.